extension/content/components/recipes/RecipeEditor.js

```diff
@@ -0,0 +1,193 @@
+import React, { useState, useEffect } from "react";
+import PropTypes from "prop-types";
+import {
+  Alert,
+  Button,
+  ButtonToolbar,
+  Form,
+  FormGroup,
+  FormControl,
+  ControlLabel,
+  HelpBlock,
+  InputPicker,
+} from "rsuite";
+import { Controlled as CodeMirror } from "react-codemirror2";
+import {
+  useSelectedEnvironmentAPI,
+  useSelectedEnvironmentAuth,
+} from "devtools/contexts/environment";
+
+export default function RecipeEditor(props) {
+  const { match } = props;
+  const api = useSelectedEnvironmentAPI();
+  const auth = useSelectedEnvironmentAuth();
+  let [data, setData] = useState({});
+  let [actions, setActions] = useState([]);
+
+  async function getActionsOptions() {
+    let res = await api.fetchActions(3);
+    let actions = res.results.map(action => ({
+      label: action.name,
+      value: action.id,
+    }));
+
+    setActions(actions);
+  }
+
+  async function getRecipe(id) {
+    let recipe = await api.fetchRecipe(id, 3);
+    const { latest_revision } = recipe;
+    latest_revision.arguments = JSON.stringify(
+      latest_revision.arguments,
+      null,
+      1,
+    );
+    latest_revision.action_id = latest_revision.action.id;
+    delete latest_revision.comment;
+    setData(latest_revision);
+  }
+
+  useEffect(() => {
+    getActionsOptions();
+    if (match.params.id) {
+      getRecipe(match.params.id);
+    }
+  }, []);
+
+  const handleChange = (key, value) => {
+    setData({ ...data, [key]: value });
+  };
+
+  const handleSubmit = () => {
+    let recipeEndpoint = "recipe/";
+    let method = "POST";
+    if (match.params.id) {
+      recipeEndpoint = `recipe/${match.params.id}/`;
+      method = "PUT";
+    }
+    const token = auth.result.accessToken;
+    const requestBody = formatRequestBody();
+    const requestSave = api.request({
+      version: 3,
+      url: recipeEndpoint,
+      extraHeaders: {
+        Authorization: "Bearer " + token,
+      },
+
+      method,
+      data: requestBody,
+    });
+    requestSave
+      .then(() => {
+        location.replace("/content.html#");
+        Alert.success("Changes Saved");
+      })
+      .catch(err => {
+        Alert.error(`An Error Occurred: ${JSON.stringify(err.data)}`);
+      });
```

Comment:
this should be a new method or a few new methods on `NormandyAPI`. Let's have a quick chat about this on Friday.